### PR TITLE
Allowing a library name to be passed to the libraryResource step

### PR DESCRIPTION
**Pull request by jmcshane**
This would allow Jenkinsfile authors to specify the library name that is included to avoid duplication across resources.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflow/libs/ResourceStep.java [hunk 1]
```diff
@@ -74,10 +85,13 @@ public DescriptorImpl() {
 
         @Override protected String run() throws Exception {
             String resource = step.resource;
+            String libraryName = step.libraryName;
             Map<String,String> contents = LibraryAdder.findResources((CpsFlowExecution) getContext().get(FlowExecution.class), resource);
-            if (contents.isEmpty()) {
+            if (libraryName != null && contents.containsKey(libraryName)) {
+                return contents.get(libraryName);
+            } else if (contents.isEmpty() || (libraryName != null && !contents.containsKey(libraryName))) {
```
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/workflow/libs/ResourceStep.java]: I think it would be better to have different error message for the `libraryName != null && !contents.containsKey(libraryName))` condition saying the resource was found, but not in the specified library.

File: src/main/java/org/jenkinsci/plugins/workflow/libs/ResourceStep.java [hunk 2]
```diff
@@ -49,6 +51,15 @@
...
         return resource;
     }
+    
+    public String getLibraryName() {
+        return libraryName;
+    }
+    
+    @DataBoundSetter
+    public void setLibraryName(String libraryName) {
+        this.libraryName = libraryName;
```
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/workflow/libs/ResourceStep.java]: I would use `= Util.fixEmptyAndTrim(libraryName)` to remove surrounding whitespace from `libraryName`.

**Comment by dwnusbaum** [comment]
I just merged #48 which introduced a merge conflict, sorry! Also, it would be great to add a `help-libraryName.html` file explaining how to use the new parameter (similarly to what was done for the encoding parameter in #48).

**Comment by jmcshane** [comment]
I should be able to get around to these comments today, I’ll make sure to rebase and resolve that conflict.

**Comment by jmcshane** [comment]
@dwnusbaum Could you check now? Thanks!

File: src/main/resources/org/jenkinsci/plugins/workflow/libs/Messages.properties [hunk 3]
```diff
@@ -23,3 +23,4 @@
 LibraryDecorator.could_not_find_any_definition_of_librari=Could not find any definition of libraries {0}
 ResourceStep.library_resource_ambiguous_among_librari=Library resource {0} ambiguous among libraries {1}
 ResourceStep.no_such_library_resource_could_be_found_=No such library resource {0} could be found.
+ResourceStep.no_matching_resource_found_in_library=Library resource {0} not found in library {1}, but was present in other library
```
Comment (dwnusbaum) [inline src/main/resources/org/jenkinsci/plugins/workflow/libs/Messages.properties]: nit: Maybe reword to: "Library resource {0} not found in library {1}, although it was present in libraries {2}" and pass `contents.keySet()` as the third argument.

File: src/main/resources/org/jenkinsci/plugins/workflow/libs/ResourceStep/help-libraryName.html [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<div>
+    The shared library name to use to locate the resource.
+    If left blank, all loaded shared libraries will be loaded.
```
Comment (dwnusbaum) [inline src/main/resources/org/jenkinsci/plugins/workflow/libs/ResourceStep/help-libraryName.html]: nit: Maybe reword to "If left blank, then this step will search for the resource in all configured shared libraries."

**Comment by jglick** [comment]
Tread carefully here. I am not convinced this is safe, which is why I did not offer such an option to begin with.

**Comment by jglick** [comment]
Also from a usage perspective this is dubious. `libraryResource` is normally called from library code. But a library does not, in general, know its own name—this is picked arbitrarily by an administrator. So you could not use this feature _portably_. It could only be used by a library written for exclusive use in a particular environment. But then why would there be duplication?

It would be helpful to have an explanation of what the original problem was which you are attempting to solve via this feature.

**Comment by caporegim** [comment]
Any progress with this pull request?

**Comment by jglick** [comment]
As of #172 most of the code in this plugin has been moved to another plugin repository so this PR must be closed. If this change is still needed, please

```bash
git clone https://github.com/jenkinsci/pipeline-groovy-lib-plugin
cd pipeline-groovy-lib-plugin
git checkout -b SELECT-BRANCH-NAME
git pull https://github.com/jmcshane/workflow-cps-global-lib-plugin master
```

resolve any merge conflicts, and file a fresh PR on the new repository. Be sure to paste a link to this old PR to enable bidirectional navigation.
